File: 10-逆向工程/02-移动端逆向/iOS砸壳与加密镜像/macho_dump.py
"""iOS 砸壳：Mach-O 的 LC_ENCRYPTION_INFO(_64) 语义与"从内存 dump 回文件"的全过程。

结构体与常量全部取自 xnu 源码 EXTERNAL_HEADERS/mach-o/loader.h：
  - mach_header_64 / segment_command_64 / encryption_info_command(_64)
  - MH_MAGIC_64 / MH_CIGAM_64 / LC_SEGMENT_64 / LC_ENCRYPTION_INFO(_64)
  - MH_DYLIB_IN_CACHE（在 dyld 共享缓存里的 dylib）
参考：https://github.com/apple-oss-distributions/xnu/blob/main/EXTERNAL_HEADERS/mach-o/loader.h
"""

import struct

# ---------- 常量（loader.h） ----------

MH_MAGIC_64 = 0xFEEDFACF          # 64 位 mach header 的 magic
MH_CIGAM_64 = 0xCFFAEDFE          # NXSwapInt(MH_MAGIC_64)：字节序被交换过

LC_SYMTAB = 0x02
LC_SEGMENT_64 = 0x19
LC_UUID = 0x1B
LC_ENCRYPTION_INFO = 0x21
LC_ENCRYPTION_INFO_64 = 0x2C

MH_EXECUTE = 0x2
MH_DYLIB = 0x6
MH_PIE = 0x200000
MH_DYLIB_IN_CACHE = 0x80000000    # 只用于 dylib：该库在 dyld 共享缓存里

# ---------- 结构体大小（按 loader.h 字段推算） ----------

HEADER_64_SIZE = 32               # 8 个 uint32
SEGMENT_64_SIZE = 72              # 4+4+16+8*4+4*4
ENCRYPTION_INFO_SIZE = 20         # 5 个 uint32
ENCRYPTION_INFO_64_SIZE = 24      # 6 个 uint32（pad 使大小为 8 的倍数）

# encryption_info_command_64 内 cryptid 的偏移：cmd + cmdsize + cryptoff + cryptsize
CRYPTID_OFFSET = 16


class MachOError(Exception):
    pass


class Segment(object):

    def __init__(self, name, vmaddr, vmsize, fileoff, filesize, cmd_off):
        self.name = name
        self.vmaddr = vmaddr
        self.vmsize = vmsize
        self.fileoff = fileoff
        self.filesize = filesize
        self.cmd_off = cmd_off

    def contains_fileoff(self, off):
        return self.fileoff <= off < self.fileoff + self.filesize

    def fileoff_to_vmaddr(self, off):
        if not self.contains_fileoff(off):
            raise MachOError("文件偏移 0x%x 不在段 %s 内" % (off, self.name))
        return self.vmaddr + (off - self.fileoff)


class MachO(object):
    """最小 Mach-O 解析：header + load commands + 段表 + 加密信息。"""

    def __init__(self, blob):
        self.blob = bytearray(blob)
        self._parse_header()
        self.segments = []
        self.crypt = None
        self.crypt_cmd_off = None
        self._parse_load_commands()

    def _u32(self, off):
        return struct.unpack_from("<I", self.blob, off)[0]

    def _u64(self, off):
        return struct.unpack_from("<Q", self.blob, off)[0]

    def _parse_header(self):
        magic = self._u32(0)
        if magic == MH_MAGIC_64:
            self.swapped = False
        elif magic == MH_CIGAM_64:
            self.swapped = True          # 字节序被交换过，本模型不再做完整交换
        else:
            raise MachOError("magic 0x%08x 不是 64 位 Mach-O" % magic)
        self.cputype = self._u32(4)
        self.cpusubtype = self._u32(8)
        self.filetype = self._u32(12)
        self.ncmds = self._u32(16)
        self.sizeofcmds = self._u32(20)
        self.flags = self._u32(24)
        self.reserved = self._u32(28)

    def _parse_load_commands(self):
        off = HEADER_64_SIZE
        for _ in range(self.ncmds):
            cmd = self._u32(off)
            cmdsize = self._u32(off + 4)
            if cmdsize < 8:
                raise MachOError("cmdsize 非法")
            if cmd == LC_SEGMENT_64:
                name = bytes(self.blob[off + 8:off + 24]).split(b"\x00")[0].decode("ascii")
                vmaddr = self._u64(off + 24)
                vmsize = self._u64(off + 32)
                fileoff = self._u64(off + 40)
                filesize = self._u64(off + 48)
                self.segments.append(Segment(name, vmaddr, vmsize, fileoff, filesize, off))
            elif cmd in (LC_ENCRYPTION_INFO, LC_ENCRYPTION_INFO_64):
                self.crypt = {
                    "cmd": cmd,
                    "cryptoff": self._u32(off + 8),
                    "cryptsize": self._u32(off + 12),
                    "cryptid": self._u32(off + 16),
                }
                self.crypt_cmd_off = off
            off += cmdsize

    # --- 查询 ---
    def segment_named(self, name):
        for s in self.segments:
            if s.name == name:
                return s
        return None

    def is_encrypted(self):
        """loader.h 原话：cryptid 为 0 表示尚未加密（即已被砸壳或本就未加密）。"""
        return bool(self.crypt and self.crypt["cryptid"] != 0)

    def in_dyld_shared_cache(self):
        return bool(self.flags & MH_DYLIB_IN_CACHE)

    def dump_plan(self):
        """算出"该从内存的哪个地址、拷多少字节、写回文件的哪个偏移"。

        cryptoff 是**文件偏移**；进程里解密后的数据位于
        vmaddr + (cryptoff - segment.fileoff)。这两者的换算是砸壳最容易写错的地方。
        """
        if not self.crypt:
            raise MachOError("没有 LC_ENCRYPTION_INFO(_64)")
        cryptoff = self.crypt["cryptoff"]
        cryptsize = self.crypt["cryptsize"]
        seg = None
        for s in self.segments:
            if s.contains_fileoff(cryptoff):
                seg = s
                break
        if seg is None:
            raise MachOError("cryptoff 0x%x 不落在任何段内" % cryptoff)
        return {"segment": seg.name, "file_offset": cryptoff, "size": cryptsize,
                "vmaddr": seg.fileoff_to_vmaddr(cryptoff)}

    # --- 砸壳动作 ---
    def apply_dump(self, memory, image_vm_base, plan=None):
        """把进程内存里已解密的字节写回文件对应区间。"""
        plan = plan or self.dump_plan()
        start_mem = plan["vmaddr"] - image_vm_base
        for i in range(plan["size"]):
            self.blob[plan["file_offset"] + i] = memory[start_mem + i]
        return plan

    def patch_cryptid(self, value=0):
        """把 cryptid 置 0，让内核/加载器不再尝试解密。"""
        if self.crypt_cmd_off is None:
            raise MachOError("没有加密命令可改")
        struct.pack_into("<I", self.blob, self.crypt_cmd_off + CRYPTID_OFFSET, value)
        self.crypt["cryptid"] = value

    def bytes(self):
        return bytes(self.blob)


def page_aligned_range(cryptoff, cryptsize, page_size):
    """加密区间落在若干页内，dump 必须覆盖**整个页**。

    page_size 由调用方给出（不同平台/机型不同），本函数不假定具体数值。
    """
    start = cryptoff - (cryptoff % page_size)
    end = cryptoff + cryptsize
    end = end + ((-end) % page_size)
    return start, end - start


def build_sample(page_size=0x1000, encrypted=True):
    """构造一个带 __TEXT 段与 LC_ENCRYPTION_INFO_64 的最小 Mach-O。"""
    text_off = 0x4000
    text_size = 0x2000
    cryptsize = 0x1000
    cryptoff = text_off + 0x1000               # __TEXT 的后半段被加密
    blob = bytearray(0x8000)
    # 段内容：前半段明文标记，后半段"密文"标记
    for i in range(text_size):
        blob[text_off + i] = 0x41 if i < 0x1000 else 0xEE
    struct.pack_into("<I", blob, 0, MH_MAGIC_64)
    struct.pack_into("<I", blob, 12, MH_EXECUTE)
    struct.pack_into("<I", blob, 16, 2)        # ncmds
    struct.pack_into("<I", blob, 20, SEGMENT_64_SIZE + ENCRYPTION_INFO_64_SIZE)
    struct.pack_into("<I", blob, 24, MH_PIE)

    off = HEADER_64_SIZE
    struct.pack_into("<II", blob, off, LC_SEGMENT_64, SEGMENT_64_SIZE)
    struct.pack_into("<16s", blob, off + 8, b"__TEXT")   # segname 是定长 16 字节
    struct.pack_into("<QQQQ", blob, off + 24, 0x100000000, text_size, text_off, text_size)
    off += SEGMENT_64_SIZE

    struct.pack_into("<IIIII I".replace(" ", ""), blob, off,
                     LC_ENCRYPTION_INFO_64, ENCRYPTION_INFO_64_SIZE,
                     cryptoff, cryptsize, 1 if encrypted else 0, 0)
    return bytes(blob), {"text_off": text_off, "cryptoff": cryptoff,
                         "cryptsize": cryptsize, "page_size": page_size}

File: 10-逆向工程/02-移动端逆向/iOS砸壳与加密镜像/macho_dump.go
// macho_dump.go — Mach-O 加密段与砸壳的 Go 实现(与 macho_dump.py 同构)。
// 结构体与常量取自 xnu 源码 EXTERNAL_HEADERS/mach-o/loader.h; 本机无 go 工具链:
// 人工审查 + 括号配平校验。
package main

import (
	"encoding/binary"
	"fmt"
)

// loader.h 常量
const (
	mhMagic64           = 0xfeedfacf
	mhCigam64           = 0xcffaedfe
	lcSegment64         = 0x19
	lcUUID              = 0x1b
	lcEncryptionInfo    = 0x21
	lcEncryptionInfo64  = 0x2c
	mhExecute           = 0x2
	mhPIE               = 0x200000
	mhDylibInCache      = 0x80000000
)

// 结构体大小
const (
	header64Size          = 32 // 8 个 uint32
	segment64Size         = 72 // 4+4+16+8*4+4*4
	encryptionInfoSize    = 20 // 5 个 uint32
	encryptionInfo64Size  = 24 // 补 pad 到 8 的倍数
	cryptIDOffset         = 16 // cmd + cmdsize + cryptoff + cryptsize
)

type machHeader64 struct {
	magic      uint32
	cputype    uint32
	cpusubtype uint32
	filetype   uint32
	ncmds      uint32
	sizeofcmds uint32
	flags      uint32
	reserved   uint32
}

type segment64 struct {
	name     string
	vmaddr   uint64
	vmsize   uint64
	fileoff  uint64
	filesize uint64
}

type cryptInfo struct {
	cryptoff  uint32
	cryptsize uint32
	cryptid   uint32
	cmdOff    int
}

func parseHeader64(buf []byte) (*machHeader64, error) {
	if len(buf) < header64Size {
		return nil, fmt.Errorf("文件短于 mach_header_64 的 32 字节")
	}
	h := &machHeader64{}
	h.magic = binary.LittleEndian.Uint32(buf[0:4])
	if h.magic != mhMagic64 && h.magic != mhCigam64 {
		return nil, fmt.Errorf("magic %#x 不是 64 位 Mach-O", h.magic)
	}
	h.cputype = binary.LittleEndian.Uint32(buf[4:8])
	h.filetype = binary.LittleEndian.Uint32(buf[12:16])
	h.ncmds = binary.LittleEndian.Uint32(buf[16:20])
	h.sizeofcmds = binary.LittleEndian.Uint32(buf[20:24])
	h.flags = binary.LittleEndian.Uint32(buf[24:28])
	return h, nil
}

// parseCommands 扫 load commands, 抽出段表与加密信息
func parseCommands(buf []byte, h *machHeader64) ([]segment64, *cryptInfo, error) {
	segs := []segment64{}
	var crypt *cryptInfo
	off := header64Size
	for i := uint32(0); i < h.ncmds; i++ {
		if off+8 > len(buf) {
			return nil, nil, fmt.Errorf("load command 越界")
		}
		cmd := binary.LittleEndian.Uint32(buf[off : off+4])
		cmdsize := binary.LittleEndian.Uint32(buf[off+4 : off+8])
		if cmdsize < 8 || off+int(cmdsize) > len(buf) {
			return nil, nil, fmt.Errorf("cmdsize 非法")
		}
		switch cmd {
		case lcSegment64:
			if cmdsize < uint32(segment64Size) {
				return nil, nil, fmt.Errorf("LC_SEGMENT_64 的 cmdsize 不足 72")
			}
			name := string(buf[off+8 : off+24])
			for i := 0; i < len(name); i++ {
				if name[i] == 0 {
					name = name[:i]
					break
				}
			}
			segs = append(segs, segment64{
				name:     name,
				vmaddr:   binary.LittleEndian.Uint64(buf[off+24 : off+32]),
				vmsize:   binary.LittleEndian.Uint64(buf[off+32 : off+40]),
				fileoff:  binary.LittleEndian.Uint64(buf[off+40 : off+48]),
				filesize: binary.LittleEndian.Uint64(buf[off+48 : off+56]),
			})
		case lcEncryptionInfo, lcEncryptionInfo64:
			crypt = &cryptInfo{
				cryptoff:  binary.LittleEndian.Uint32(buf[off+8 : off+12]),
				cryptsize: binary.LittleEndian.Uint32(buf[off+12 : off+16]),
				cryptid:   binary.LittleEndian.Uint32(buf[off+16 : off+20]),
				cmdOff:    off,
			}
		}
		off += int(cmdsize)
	}
	return segs, crypt, nil
}

// isEncrypted: loader.h 原文 —— cryptid 为 0 表示尚未加密
func isEncrypted(c *cryptInfo) bool { return c != nil && c.cryptid != 0 }

// dumpPlan: cryptoff 是**文件偏移**, 内存地址要用所在段换算
type dumpPlan struct {
	segment    string
	fileOffset uint32
	size       uint32
	vmaddr     uint64
}

func makeDumpPlan(segs []segment64, c *cryptInfo) (*dumpPlan, error) {
	if c == nil {
		return nil, fmt.Errorf("没有 LC_ENCRYPTION_INFO(_64)")
	}
	for _, s := range segs {
		if uint64(c.cryptoff) >= s.fileoff && uint64(c.cryptoff) < s.fileoff+s.filesize {
			return &dumpPlan{
				segment:    s.name,
				fileOffset: c.cryptoff,
				size:       c.cryptsize,
				vmaddr:     s.vmaddr + (uint64(c.cryptoff) - s.fileoff),
			}, nil
		}
	}
	return nil, fmt.Errorf("cryptoff %#x 不落在任何段内", c.cryptoff)
}

// applyDump 把进程内存里已解密的字节写回文件区间
func applyDump(file []byte, memory []byte, imageBase uint64, p *dumpPlan) {
	startMem := p.vmaddr - imageBase
	for i := uint32(0); i < p.size; i++ {
		file[int(p.fileOffset)+int(i)] = memory[int(startMem)+int(i)]
	}
}

// patchCryptID 把 cryptid 置 0
func patchCryptID(file []byte, c *cryptInfo) {
	binary.LittleEndian.PutUint32(file[c.cmdOff+cryptIDOffset:], 0)
	c.cryptid = 0
}

// pageAlignedRange: 加密区间落在若干页内, dump 必须覆盖整个页。
// 页大小由调用方给出(不同平台不同), 这里不假定具体数值。
func pageAlignedRange(cryptoff, cryptsize, pageSize uint64) (uint64, uint64) {
	start := cryptoff - cryptoff%pageSize
	end := cryptoff + cryptsize
	if end%pageSize != 0 {
		end += pageSize - end%pageSize
	}
	return start, end - start
}

func main() { selfcheck() }
